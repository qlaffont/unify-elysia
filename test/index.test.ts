/* eslint-disable @typescript-eslint/ban-ts-comment */
import { treaty } from '@elysiajs/eden';
import { beforeAll, describe, expect, it } from 'bun:test';
import Elysia from 'elysia';

import { app } from './utils/server.test';

const testRoute = async (
  server: Elysia,
  routePath: string,
  supposedMessage: Record<string, unknown>,
  supposedStatus: number,
): Promise<unknown> => {
  let status: number;

  const JSONResponse = await server
    .handle(new Request(`http://localhost${routePath}`))
    .then((res) => {
      status = res.status;
      return res.json();
    });

  expect(JSONResponse).toMatchObject(supposedMessage);
  expect(status!).toBe(supposedStatus);

  return JSONResponse;
};

describe('Unify Elysia', () => {
  let api: ReturnType<typeof treaty<ReturnType<typeof app>>>;
  let currentApp: Elysia;

  beforeAll(() => {
    //@ts-ignore
    currentApp = app();
    //@ts-ignore
    api = treaty(currentApp);
  });

  it('should handle validation', async () => {
    // @ts-ignore
    const { error, status } = await api.validation.get({});

    expect(status).toMatchSnapshot();
    expect(JSON.stringify(error)).toMatchSnapshot();
  });

  it('should handle elysia error', async () => {
    // @ts-ignore
    const { error, status } = await api['elysia-error'].get({});

    expect(status).toMatchSnapshot();
    expect(JSON.stringify(error)).toMatchSnapshot();
  });

  it('should handle generic error', async () => {
    // @ts-ignore
    const { error, status } = await api['generic-error'].get({});

    expect(status).toMatchSnapshot();
    expect(JSON.stringify(error)).toMatchSnapshot();
  });

  it('should handle error from unify', async () => {
    // @ts-ignore
    const { error, status } = await api['unify-error'].get({});

    expect(status).toMatchSnapshot();
    expect(JSON.stringify(error)).toMatchSnapshot();
  });

  it('bad request', async () => {
    await testRoute(
      currentApp,
      '/bad-request',
      {
        error: 'Bad Request',
        context: { example: 'A bad request error' },
      },
      400,
    );
  });

  it('unauthorized', async () => {
    await testRoute(
      currentApp,
      '/unauthorized',
      {
        error: 'Unauthorized',
        context: { example: 'An unauthorized error' },
      },
      401,
    );
  });

  it('forbidden', async () => {
    await testRoute(
      currentApp,
      '/forbidden',
      {
        error: 'Forbidden',
        context: { example: 'A forbidden error' },
      },
      403,
    );
  });

  it('not-found', async () => {
    await testRoute(
      currentApp,
      '/not-found',
      {
        error: 'Not Found',
        context: { example: 'A not found error' },
      },
      404,
    );

    await testRoute(
      currentApp,
      '/not-found-url-not-registered',
      {
        error: 'Not Found',
      },
      404,
    );
  });

  it('time-out', async () => {
    await testRoute(
      currentApp,
      '/request-time-out',
      {
        error: 'Request Time-out',
        context: { example: 'A request time out error' },
      },
      408,
    );
  });

  it('internal', async () => {
    await testRoute(
      currentApp,
      '/internal',
      {
        error: 'Internal Server Error',
        context: { example: 'An internal server error' },
      },
      500,
    );
  });

  it('not-implemented', async () => {
    await testRoute(
      currentApp,
      '/not-implemented',
      {
        error: 'Not Implemented',
        context: { example: 'A not implemented error' },
      },
      501,
    );
  });

  it('rate limit', async () => {
    await testRoute(
      currentApp,
      '/rate-limit',
      {
        error: 'Too Many Requests',
      },
      429,
    );
  });

  it("error not extending 'CustomError' from 'unify-errors'", async () => {
    await testRoute(
      currentApp,
      '/not-custom',
      { error: 'An unexpected error occured' },
      500,
    );
  });

  it("default error extending 'CustomError' from 'unify-errors'", async () => {
    await testRoute(
      currentApp,
      '/default-case',
      {
        error: 'An unexpected error occured',
        context: { example: 'A CustomError but not handled' },
        errorDetails: 'A default error',
      },
      500,
    );
  });
});
