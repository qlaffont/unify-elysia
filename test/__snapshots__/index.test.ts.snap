// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Unify Elysia should handle validation 1`] = `400`;

exports[`Unify Elysia should handle validation 2`] = `"{"status":400,"value":{"error":"Bad Request - validation","context":"Required property","stack":"Error: {\n  \"type\": \"validation\",\n  \"on\": \"query\",\n  \"property\": \"/name\",\n  \"message\": \"Required property\",\n  \"expected\": {\n    \"name\": \"\"\n  },\n  \"found\": {},\n  \"errors\": [\n    {\n      \"type\": 45,\n      \"schema\": {\n        \"type\": \"string\"\n      },\n      \"path\": \"/name\",\n      \"message\": \"Required property\"\n    },\n    {\n      \"type\": 54,\n      \"schema\": {\n        \"type\": \"string\"\n      },\n      \"path\": \"/name\",\n      \"message\": \"Expected string\"\n    }\n  ]\n}\n    at new _ValidationError (/home/quentin/Documents/dev/unify-elysia/node_modules/elysia/dist/index.mjs:2090:10)\n    at <anonymous> (:72:50)\n    at processTicksAndRejections (:12:39)","on":"query","expected":{"name":""},"errors":[{"type":45,"schema":{"type":"string"},"path":"/name","message":"Required property"},{"type":54,"schema":{"type":"string"},"path":"/name","message":"Expected string"}]}}"`;

exports[`Unify Elysia should handle elysia error 1`] = `401`;

exports[`Unify Elysia should handle elysia error 2`] = `"{"status":401,"value":"Unauthorized"}"`;

exports[`Unify Elysia should handle generic error 1`] = `500`;

exports[`Unify Elysia should handle generic error 2`] = `"{"status":500,"value":{"error":"An unexpected error occured"}}"`;

exports[`Unify Elysia should handle error from unify 1`] = `400`;

exports[`Unify Elysia should handle error from unify 2`] = `"{"status":400,"value":{"error":"Bad Request","stack":"Error: Bad Request\n    at construct (native)\n    at _construct (/home/quentin/Documents/dev/unify-elysia/node_modules/unify-errors/dist/unify-errors.cjs.development.js:4:37)\n    at Wrapper (/home/quentin/Documents/dev/unify-elysia/node_modules/unify-errors/dist/unify-errors.cjs.development.js:59:26)\n    at CustomError (/home/quentin/Documents/dev/unify-elysia/node_modules/unify-errors/dist/unify-errors.cjs.development.js:77:26)\n    at new BadRequest (/home/quentin/Documents/dev/unify-elysia/node_modules/unify-errors/dist/unify-errors.cjs.development.js:90:32)\n    at <anonymous> (/home/quentin/Documents/dev/unify-elysia/test/utils/server.test.ts:35:25)\n    at <anonymous> (:71:16)\n    at processTicksAndRejections (:12:39)","errorDetails":"Bad Request"}}"`;
